File: backend/domains/gateway/presentation/schemas/resource_grants.py
"""个人资源 grant API schemas。"""

from __future__ import annotations

import uuid

from pydantic import BaseModel, Field


class ResourceGrantCreateRequest(BaseModel):
    subject_kind: str = Field(..., pattern="^(credential|model)$")
    subject_id: uuid.UUID
    target_team_ids: list[uuid.UUID] = Field(..., min_length=1)
    note: str | None = None


class ResourceGrantUpdateRequest(BaseModel):
    enabled: bool | None = None
    note: str | None = None


class ResourceGrantResponse(BaseModel):
    id: uuid.UUID
    owner_user_id: uuid.UUID
    subject_kind: str
    subject_id: uuid.UUID
    target_team_id: uuid.UUID
    enabled: bool
    note: str | None
    granted_by: uuid.UUID

    model_config = {"from_attributes": True}


class GrantedModelResponse(BaseModel):
    model_id: uuid.UUID
    name: str
    real_model: str
    provider: str
    capability: str
    credential_id: uuid.UUID
    owner_user_id: uuid.UUID
    personal_team_id: uuid.UUID


__all__ = [
    "GrantedModelResponse",
    "ResourceGrantCreateRequest",
    "ResourceGrantResponse",
    "ResourceGrantUpdateRequest",
]

File: frontend/src/api/gateway/resource-grants.ts
import { apiClient } from '@/api/client'

import { GATEWAY_API_BASE, teamGatewayPath } from './_base'

export interface ResourceGrant {
  id: string
  owner_user_id: string
  subject_kind: 'credential' | 'model'
  subject_id: string
  target_team_id: string
  enabled: boolean
  note: string | null
  granted_by: string
}

export interface ResourceGrantCreateBody {
  subject_kind: 'credential' | 'model'
  subject_id: string
  target_team_ids: string[]
  note?: string | null
}

export interface GrantedModel {
  model_id: string
  name: string
  real_model: string
  provider: string
  capability: string
  credential_id: string
  owner_user_id: string
  personal_team_id: string
}

export const resourceGrantsApi = {
  listMine: () => apiClient.get<ResourceGrant[]>(`${GATEWAY_API_BASE}/resource-grants`),

  create: (body: ResourceGrantCreateBody) =>
    apiClient.post<ResourceGrant[]>(`${GATEWAY_API_BASE}/resource-grants`, body),

  revoke: (grantId: string) =>
    apiClient.delete<unknown>(`${GATEWAY_API_BASE}/resource-grants/${grantId}`),

  listTeamGrantedModels: (teamId: string) =>
    apiClient.get<GrantedModel[]>(teamGatewayPath(teamId, '/granted-resources/models')),
}
